### PR TITLE
Disallow pointer types to be used as top-level constraints on generic parameters

### LINKED ISSUE #40182: Compiler incorrectly compiling pointer constraints
Consider this code

```cs
unsafe class Generic<T> where T : byte*
{
}
```

This should issue the following error:

> Program.cs(10,39): error CS0706: Invalid constraint type. A type used as a constraint must be an interface, a non-sealed class or        a type parameter.

However the code is instead compiling. 

Checked with the native csc compiler and it does issue this error.

### PR DESCRIPTION
**Pull request by 333fred**
Fixes #40182

Tagging @dotnet/roslyn-compiler @jaredpar for review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs [hunk 1]
```diff
@@ -223,13 +223,12 @@ internal partial class Binder
                             var typeSyntax = typeConstraintSyntax.Type;
                             var typeSyntaxKind = typeSyntax.Kind();
 
-                            // For pointer types, don't report this error. It is already reported during binding typeSyntax below.
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs]: Was this a regression?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs]: Compared to the native compiler? Yes. Compared to initial versions of roslyn? I'm not certain.

---
In reply to: [354631451](https://github.com/dotnet/roslyn/pull/40195#discussion_r354631451) [](ancestors = 354631451)

File: src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs [hunk 2]
```diff
@@ -223,13 +223,12 @@ internal partial class Binder
...
                             switch (typeSyntaxKind)
                             {
                                 case SyntaxKind.PredefinedType:
-                                case SyntaxKind.PointerType:
                                 case SyntaxKind.NullableType:
                                     break;
+                                case SyntaxKind.PointerType:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs]: >case SyntaxKind.PointerType: [](start = 32, length = 28)

I am not sure why are we splitting reporting invalid constraint types between this place and Binder.IsValidConstraintType. Moreover, this place does it based on the syntax, which would still leave a hole if we could have aliases for pointers and array types. If in the future we relax the alias restriction, we will get in trouble. Consider consolidating the checks to one place and performing it based on symbol.

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

**Pull request merged**